File: p22/compound.m
function Xwb = compound(Xwa, Xab)
% 
% For details, please refer to Section 3.2.1 of R. Smith, M. Self, 
% "Estimating uncertain spatial relationships in robotics," 
% Autonomous Robot Vehicles, pp. 167-193, 1990. 
%   Xwa - should be of size 3*1 or 1*3
%   Xab - can be 3*n or 2*n matrix
% Seeded by Tim Bailey
% Modified by Samuel on 7 May 2013

   rot = [cos(Xwa(3))    -sin(Xwa(3)); 
          sin(Xwa(3))     cos(Xwa(3))];
   
   Xwb(1:2,:) = rot * Xab(1:2,:) + repmat(Xwa(1:2), 1, size(Xab,2));

    %% if Xwb is a pose and not a point
    if size(Xab,1) == 3
       Xwb(3,:) = piTopi(Xwa(3,:) + Xab(3));
    end
end


function angle = piTopi(angle)
    % Input: array of angles.
    % Output: normalised angles.
    % Tim Bailey 2000, modified 2005.

    % Note: either rem or mod will work correctly
    %angle = mod(angle, 2*pi); % mod() is very slow for some versions of MatLab (not a builtin function)
    %angle = rem(angle, 2*pi); % rem() is typically builtin

    twopi = 2*pi;
    angle = angle - twopi*fix(angle/twopi); % this is a stripped-down version of rem(angle, 2*pi)

    i = find(angle >= pi);
    angle(i) = angle(i) - twopi;

    i = find(angle < -pi);
    angle(i) = angle(i) + twopi;
end

File: Codes.python/p34/compound.py
import numpy as np
from math import*
def compound(Xwa,Xab):
    Xwb = np.array(np.zeros([3,3]))
    rot = np.array([[cos(Xwa[2,0]),-sin(Xwa[2,0] )],
                  [sin(Xwa[2,0]), cos(Xwa[2,0])]])
    Xwb[0:2] = rot.dot(np.array([Xab[0],Xab[1]])) + np.tile([Xwa[0:2]],(1, len(Xab[0])))
    row = np.shape(Xab)
    if row[0] == 3:
        Xwb[2] = piTopi(Xwa[2] + Xab[2])
    return Xwb

def piTopi(angle):
    twopi = 2*pi
    angle = angle - twopi * int(angle/twopi)
    i1 = np.where(angle >= pi)
    angle[i1] = angle[i1] - twopi

    i2 = np.where(angle < -pi)
    angle[i2] = angle[i2] + twopi
    return angle

# a =np.array([[5],[3],[5*pi/18]])
# b =np.array([[4],[2],[pi/9]])
# print compound(a,b)